sycl/include/sycl/info/kernel_traits.def

```diff
@@ -1,5 +1,6 @@
-__SYCL_PARAM_TRAITS_SPEC(kernel, num_args, uint32_t, PI_KERNEL_INFO_NUM_ARGS)
-__SYCL_PARAM_TRAITS_SPEC(kernel, attributes, std::string, PI_KERNEL_INFO_ATTRIBUTES)
-__SYCL_PARAM_TRAITS_SPEC(kernel, function_name, std::string, PI_KERNEL_INFO_FUNCTION_NAME)
-__SYCL_PARAM_TRAITS_SPEC(kernel, reference_count, uint32_t, PI_KERNEL_INFO_REFERENCE_COUNT)
-__SYCL_PARAM_TRAITS_SPEC(kernel, context, sycl::context, PI_KERNEL_INFO_CONTEXT)
+__SYCL_PARAM_TRAITS_SPEC(kernel, num_args, uint32_t, UR_KERNEL_INFO_NUM_ARGS)
+__SYCL_PARAM_TRAITS_SPEC(kernel, attributes, std::string, UR_KERNEL_INFO_ATTRIBUTES)
+__SYCL_PARAM_TRAITS_SPEC(kernel, function_name, std::string, UR_KERNEL_INFO_FUNCTION_NAME)
+__SYCL_PARAM_TRAITS_SPEC(kernel, reference_count, uint32_t, UR_KERNEL_INFO_REFERENCE_COUNT)
+__SYCL_PARAM_TRAITS_SPEC(kernel, context, sycl::context, UR_KERNEL_INFO_CONTEXT)
+__SYCL_PARAM_TRAITS_SPEC(kernel, ext_codeplay_num_regs, uint32_t, UR_KERNEL_INFO_NUM_REGS)
```

Comment:
where this info was located before? are we adding something new here?

Reply:
This was moved here from [kernel_device_specific_traits.def](https://github.com/intel/llvm/pull/14145/files#diff-bf22d9e096451ba319209b687d27b264f3d3fc6073f16a80ca9076c804b57eba) because in PI `NUM_REGS` was a `GROUP_INFO` enum, whereas in UR it's a regular `KERNEL_INFO` enum so it needs its info helper to be instantiated in here instead because of how the .def files are included. If this causes non-obvious problems I think we could make some changes to `info_desc_helpers.hpp` instead

Reply:
Few concerns here:
- This particular change causes an ABI-breaking change (which I would like to avoid)
- This likely also causes API change without any deprecation messages which I think is disallowed even within ABI-breaking window
- It seems like we have no tests for this functionality at all and no language specification. For me that's a candidate for just being removed until we have a formal spec for it. I really don't want our customers to us an undocumented interface

Reply:
I have a patch that allows us to revert this change, it makes the aforementioned helper header a bit nastier by introducing a nested `std::conditional` but it's a relatively minor price to pay

Reply:
> I have a patch that allows us to revert this change, it makes the aforementioned helper header a bit nastier by introducing a nested `std::conditional` but it's a relatively minor price to pay

Haven't looked into the current issue, but "ugly nested conditional" sounds similar enough to https://github.com/intel/llvm/blob/0360e6af2a353210d508633a60ff02327094f7e7/sycl/include/sycl/detail/type_traits.hpp#L553-L565, maybe something like this can help here as well.

Reply:
I did look into it but I don't think that's applicable here unfortunately: it isn't a mapping of type -> type, it's a mapping of enum value -> type. See [here](https://github.com/intel/llvm/pull/14145/files#diff-fc49f47ad53d5714302312c21e513c99f8ea464799f05837779fdae9bc3534a1R94)